lightning/src/ln/channelmanager.rs

```diff
@@ -2905,16 +2920,30 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						// event being update_fulfill_htlc).
 						if let Err(e) = self.chain_monitor.update_channel(prev_outpoint, preimage_update) {
 							log_error!(self.logger, "Critical error: failed to update channel monitor with preimage {:?}: {:?}",
-							           payment_preimage, e);
+									   payment_preimage, e);
 						}
 						Ok(())
 					},
 					Err(Some(res)) => Err(res),
-				} {
-					mem::drop(channel_state_lock);
+				};
+				mem::drop(channel_state_lock);
+				if let Err((counterparty_node_id, err)) = res {
 					let res: Result<(), _> = Err(err);
 					let _ = handle_error!(self, res, counterparty_node_id);
 				}
+				if claimed_htlc {
+					if let Some(forwarded_htlc_value) = forwarded_htlc_value_msat {
+						let fee_earned_msat = if let Some(claimed_htlc_value) = claimed_value_msat {
+							Some(claimed_htlc_value - forwarded_htlc_value)
+						} else { None };
+
+						let mut pending_events = self.pending_events.lock().unwrap();
+						pending_events.push(events::Event::PaymentForwarded {
+							fee_earned_msat,
+							claim_from_onchain_tx: from_onchain,
```

Comment:
Will this be generated even if we're `claiming_funds` for a payment that was directly to our node?

Reply:
Hmm, I think not? We should never get a `claim_funds_internal` call from an HTLC to us. That should come in via (the public) `claim_funds` method.